xla/service/gpu/transforms/cudnn_fused_conv_rewriter.cc

```diff
@@ -488,83 +564,138 @@ void CaptureConvGraphRecursive(HloInstruction* instr,
+  // followed by a recursive call of CaptureConvGraphRecursive to match and
+  // potentially fuse its users. Aside from scalar reduction results, the graph
+  // must have a single tensor endpoint.
+  //
+  // Examples:
+  //
+  //        C
+  //       / \
+  //  A - B   E - F
+  //       \ /
+  //        D
+  //
+  // The graph is fully fused since it has one endpoint.
+  //
+  //        C - D
+  //       /
+  //  A - B
+  //       \
+  //        E - F
+  //
+  // Fusion stops at B since the graph would have two endpoints.
+  int num_new_users = 0;
+  int num_existing_users = 0;
   for (HloInstruction* user : instr->users()) {
-    HloInstruction *op, *operand0, *operand1;
+    HloInstruction *op0, *op1, *op2, *operand0, *operand1;
+    UniqueHloInstruction unique_operand;
     // Add
-    if (Match(user, m::AddAnyOrder(&op, m::Op(&operand0), m::Op(&operand1)))) {
-      if (graph_string.AppendOp("add", op, {operand0, operand1})) {
+    if (Match(user, m::AddAnyOrder(&op0, m::Op(&operand0), m::Op(&operand1)))) {
+      if (graph_string.AppendOp("add", op0, {operand0, operand1})) {
         operands.push_back(operand0 == instr ? operand1 : operand0);
-        num_linear_users++;
+        ++num_new_users;
         CaptureConvGraphRecursive(user, operands, aux_outputs, graph_string,
-                                  visited_instrs, final_instr);
+                                  visited_instrs, final_instr, num_endpoints);
+      } else {
+        // Since operands only holds ops that are not part of the graph, remove
+        // instr.
+        operands.erase(std::remove(operands.begin(), operands.end(), instr),
+                       operands.end());
+        ++num_existing_users;
       }
       continue;
     }
     // Scale
-    if (Match(user, m::MultiplyAnyOrder(&op, m::Op(&operand0),
+    if (Match(user, m::MultiplyAnyOrder(&op0, m::Op(&operand0),
                                         m::Broadcast(m::Op(&operand1)))) &&
         ShapeUtil::IsScalar(operand1->shape())) {
-      if (graph_string.AppendOp("scale", op, {operand0, operand1})) {
+      if (graph_string.AppendOp("scale", op0, {operand0, operand1})) {
         operands.push_back(operand1);
-        num_linear_users++;
+        ++num_new_users;
         CaptureConvGraphRecursive(user, operands, aux_outputs, graph_string,
-                                  visited_instrs, final_instr);
+                                  visited_instrs, final_instr, num_endpoints);
       }
       continue;
     }
     // Inverse Scale
-    if (Match(user, m::Divide(&op, m::Op(&operand0),
+    if (Match(user, m::Divide(&op0, m::Op(&operand0),
                               m::Broadcast(m::Op(&operand1)))) &&
         ShapeUtil::IsScalar(operand1->shape())) {
-      if (graph_string.AppendOp("invscale", op, {operand0, operand1})) {
+      if (graph_string.AppendOp("invscale", op0, {operand0, operand1})) {
         operands.push_back(operand1);
```

Comment:
For Scale/InvScale, it's worth mentioning (and even DCHECKing (and even CHECKing)) that instr is operand0, not operand1. Since outputs of non-amax nodes in the graph are always non-scalar, and operand1 is scalar.